### ISSUE TITLE
Queries string

### ISSUE THREAD

**Issue by 0l4y4** [openNDS/openNDS#352]
Hello again,

I would like to know all queries string (level 3 FAS). I know the following:

- GET (fas, iv) 
- GET (clientip, gatewayname, status)
- POST (auth_get)
- POST (payload)
- POST (gatewayhash)

Any more?

Thanks!

**Comment by bluewavenet** [comment]
@0l4y4 
The only query string originating from openNDS is in the url the client is redirected to, ie GET(fas, iv).
The query string a client sees when communicating with the FAS must also contain at least GET(fas, iv), but the FAS can add any number of parameters/variables it likes during the FAS-Client dialogue.

eg .The fas-aes-https example script adds fullname and email variables during the dialogue.

**Comment by 0l4y4** [comment]
@bluewavenet 

And this?

> If the user of an already authenticated client device manually accesses the NDS Virtual URL, they will be redirected back to FAS with the “status” query string.
> 
> This will be of the form:
> 
> http://fasremoteip:fasport/faspath?clientip=[clientip]&gatewayname=[gatewayname]&status=authenticated_

On the other hand, authmon sends POST queries string, no?

In the end, what I need to know is what type of requests and with what parameters can they reach me on my server from the router

**Comment by bluewavenet** [comment]
@0l4y4 

> If the user of an already authenticated client device manually accesses the NDS Virtual URL

A client on a level 3 FAS will never be given the virtual url to openNDS (because level 3 does not use virtual urls), so this will not happen.

> authmon sends POST queries string

Yes it does, but this has nothing to do with client/FAS interaction, rather it is the authmon daemon facilitating inbound NAT traversal.

**Comment by 0l4y4** [comment]
@bluewavenet 

What I want to know is what kind of requests are going to reach my FAS. Actually I don't care if they come from the client or the router

**Comment by 0l4y4** [comment]
@bluewavenet 

On the other hand, you tell me that level 3 doesn't work with GET (clientip, gatewayname, status), then,  why fas-aes-https.php has a view called status_page() and also this code (for example)?

```
// Get the query string components
if (isset($_GET['status'])) {
	@$redir=$_GET['redir'];
	@$redir_r=explode("fas=", $redir);
	@$fas=$redir_r[1];

	if (isset($_GET['iv'])) {
		$iv=$_GET['iv'];
	} else {
		$iv="error";
	}
```

**Comment by bluewavenet** [comment]
@0l4y4 

> you tell me that level 3 doesn't work with GET

I did not tell you this. What I did say was:

> A client on a level 3 FAS will never be given the virtual url to openNDS

On an https FAS (level 3), if a client was given an openNDS virtual url, then the client would receive a dire security error.
So the level 0, 1 and 2 virtual urls cannot be used. This is by design to enforce encrypted communications. Instead, FAS communicates directly with openNDS using the encrypted NAT traversal channel opened by authmon.

> What I want to know is what kind of requests are going to reach my FAS

To clarify, when accessing a remote https FAS:

- Clients use GET requests
- Authmon uses POST requests

1. The initial redirect url sent by openNDS
  The initial redirect url sent by openNDS to clients contains fas and iv variables.
  In the rare and special case of a client authenticated via their desktop browser (eg Chrome, Firefox etc) and not the CPD "built   in" of the operating system, if the authenticated client clicks back on their browser to the initial login page then fills it in again, they will be sent the url containing fas, iv and status.

2. Subsequent communication is between the client and the remote https FAS.
  The url will contain **_at least_** fas and iv variables, **_along with any number of FAS defined additional variables_** (eg Fullname, Email).
